crates/shared/src/price_estimation/competition.rs

```diff
@@ -0,0 +1,148 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::{anyhow, Result};
+use futures::future;
+
+/// Price estimator that pulls estimates from various sources
+/// and competes on the best price.
+pub struct CompetitionPriceEstimator {
+    inner: Vec<(String, Box<dyn PriceEstimating>)>,
+}
+
+impl CompetitionPriceEstimator {
+    pub fn new(inner: Vec<(String, Box<dyn PriceEstimating>)>) -> Self {
+        assert!(!inner.is_empty());
+        Self { inner }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for CompetitionPriceEstimator {
+    async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>> {
+        let all_estimates =
+            future::join_all(self.inner.iter().map(|(name, estimator)| async move {
```

Comment:
I assume these are already all timeouted because they use our HTTP Client? Maybe add as a consideration (not sure if adding its own timeout here is necessary)?

Reply:
I'll double check. Ideally we should have a `TimeBoundPriceEstimator` for estimators that don't timeout because of IO.